### PR TITLE
Add target_float

### PR DESCRIPTION
**Pull request by Zoxc**
Adds target_float as suggested in https://github.com/rust-lang/rfcs/issues/1364

### CODE REVIEW

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -44,7 +44,9 @@ pub struct Wrapping<T>(#[stable(feature = "rust1", since = "1.0.0")] pub T);
 pub mod wrapping;
 
 // All these modules are technically private and only exposed for libcoretest:
+#[cfg_attr(not(stage0), cfg(target_float))]
 pub mod flt2dec;
+#[cfg_attr(not(stage0), cfg(target_float))]
 pub mod dec2flt;
 pub mod bignum;
 pub mod diy_float;
```
Comment (rkruppe) [inline src/libcore/num/mod.rs]: If flt2dec and dec2flt are cfg'd out, then bignum and diy_float can also be cfg'd out..

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -44,7 +44,9 @@ pub struct Wrapping<T>(#[stable(feature = "rust1", since = "1.0.0")] pub T);
 pub mod wrapping;
 
 // All these modules are technically private and only exposed for libcoretest:
```
Comment (rkruppe) [inline src/libcore/num/mod.rs]: It may be a good idea to also cfg' out the corresponding tests, otherwise those will fail.

Comment (rkruppe) [inline src/libcore/num/mod.rs]: Same might hold for the many tests that deal with math functions.

**Comment by Zoxc** [comment]
@rkruppe Tests require libtest which require librand and libstd which require floats, so that won't be very beneficial until libstd can build without floats.

**Comment by rkruppe** [comment]
Oh, right, sorry!

**Comment by alexcrichton** [comment]
The libs team discussed this during triage yesterday, and we reached a few conclusions

* There... are many overlapping concerns on this PR. It's ended up being unclear that this was even a libs team PR.
* This PR adds support for a `target_float` `#[cfg]` directive defined in custom target specs, but it only goes halfway. The language still defines the types and they're still available for use.
* In terms of an API, we likely want to consider this in conjunction with architecture-dependent types. For example `AtomicU64` might also be an architecture-dependent type. We've long wanted a mechanism for this, but haven't yet designed one.

For now we concluded that this PR may be a bit premature. There are enough concerns here (and this is touching core-enough types) that this likely wants to go through an RFC before a PR. Thanks regardless though for the PR @Zoxc!

**Pull request closed without merging**